internal/store/verticalpodautoscaler.go

```diff
@@ -0,0 +1,256 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package store
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/kube-state-metrics/pkg/constant"
+
+	autoscaling "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+	vpaclientset "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/client/clientset/versioned"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kube-state-metrics/pkg/metric"
+)
+
+var (
+	descVerticalPodAutoscalerLabelsName          = "kube_verticalpodautoscaler_labels"
+	descVerticalPodAutoscalerLabelsHelp          = "Kubernetes labels converted to Prometheus labels."
+	descVerticalPodAutoscalerLabelsDefaultLabels = []string{"namespace", "verticalpodautoscaler", "target_api_version", "target_kind", "target_name"}
+
+	vpaMetricFamilies = []metric.FamilyGenerator{
+		{
+			Name: descVerticalPodAutoscalerLabelsName,
+			Type: metric.Gauge,
+			Help: descVerticalPodAutoscalerLabelsHelp,
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				labelKeys, labelValues := kubeLabelsToPrometheusLabels(a.Labels)
+				return &metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   labelKeys,
+							LabelValues: labelValues,
+							Value:       1,
+						},
+					},
+				}
+			}),
+		},
+		{
+			Name: "kube_verticalpodautoscaler_update_mode",
+			Type: metric.Gauge,
+			Help: "Update mode of the VerticalPodAutoscaler.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+
+				if a.Spec.UpdatePolicy.UpdateMode == nil {
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+
+				for _, mode := range []autoscaling.UpdateMode{
+					autoscaling.UpdateModeOff,
+					autoscaling.UpdateModeInitial,
+					autoscaling.UpdateModeRecreate,
+					autoscaling.UpdateModeAuto,
+				} {
+					var v float64
+					if *a.Spec.UpdatePolicy.UpdateMode == mode {
+						v = 1
+					} else {
+						v = 0
+					}
+					ms = append(ms, &metric.Metric{
+						LabelKeys:   []string{"update_mode"},
+						LabelValues: []string{string(mode)},
+						Value:       v,
+					})
+				}
+
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_verticalpodautoscaler_container_resource_policy_min",
```

Comment:
we should stay as close as possible to the original naming, so this should be `kube_verticalpodautoscaler_resourcepolicy_containerpolicies`.